### PR TITLE
Refactor traefik with package

### PR DESCRIPTION
**Pull request by vdemeester**
Split a bit traefik into package. The idea behind this refactor is to start move inter-dependencies away and do some DRY or SRP or both :stuck_out_tongue_closed_eyes: . 🐙

- Adds a `provider` package, with providers except `web.go`
- Adds a `types` package with common struct.
- Move `gen.go` to an `autogen` package

The 2nd commit is *optional*, but will be needed at some point :see_no_evil:.
To test heavily as it might breaks things that are not covered yet :smile_cat:.

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by ldez** [comment]
dry-clean-led-sign-l3000

**Comment by emilevauge** [comment]
Great!
Just to understand, `web.go` needs more refactoring before moving it ?

**Comment by vdemeester** [comment]
@emilevauge yeah, but I was even wondering if it should treat the same as the other providers — because to me, it's not the same kind of *provider* (per se). 

That's why I did not move it to the `provider` package.

**Comment by emilevauge** [comment]
OK, I understand it's a bit special. @ldez, what do you think? Moving it in `provider` package or not?

**Comment by ldez** [comment]
I agree with @vdemeester : `WebProvider` is different.
The API part can be assimilate as other provider but static content not.

It's contains different kind of things : REST API & static content.
Using the web but not with the same purpose.

Review the name ?

**Comment by vdemeester** [comment]
@ldez I would even handle it differently (and not making it implement `Provider`).

**Comment by emilevauge** [comment]
That's true. I know it should be done differently. We should make a proposal.

**Comment by vdemeester** [comment]
Ok, passing this to code review then :wink:. We'll refactor the `web` provider in another PR (to make this one not too big).

File: generate.go [hunk 1]
```diff
@@ -3,7 +3,7 @@ Copyright
 */
 
 //go:generate rm -vf gen.go
```
Comment (emilevauge) [inline generate.go]: Better with `//go:generate rm -vf autogen/gen.go` no ?

**Comment by emilevauge** [comment]
Only one little comment, other than that, LGTM!
Well done @vdemeester :)

File: generate.go (hunk 1 above)
Comment (vdemeester) [inline generate.go]: :sweat_smile: maybe ? :facepalm:
facepalm5

**Pull request merged**